### PR TITLE
fix: add messageId, rudderId and type based on new schema

### PR DESCRIPTION
**Pull request by BonapartePC**
# Description

add messageId, rudderId and request type based on new rudder schema in the internal batch endpoint.

Related PRs:
https://github.com/rudderlabs/rudder-schemas/pull/35
https://github.com/rudderlabs/rudder-ingestion-svc/pull/234
https://github.com/rudderlabs/rudder-src-router/pull/144

## Linear Ticket

Fixes PIPE-1557

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Review by Sidddddarth — APPROVED** [review]

File: gateway/gateway_test.go [hunk 1]
```diff
@@ -2241,6 +2217,141 @@ var _ = Describe("Gateway", func() {
...
+			}
+			msg := stream.Message{
+				Properties: properties,
+				Payload:    []byte(`{}`),
+			}
+			messages := []stream.Message{msg}
+			payload, err := json.Marshal(messages)
+			Expect(err).To(BeNil())
+			req := &webRequestT{
+				reqType:        "batch",
```
Comment (fracasula) [inline gateway/gateway_test.go:2331]: I searched in the whole file and this one is always `batch`. We are we testing that we don't add `"batch"` for other request types like `track`?

In the ingestion-svc we are not just setting the request type, we are also doing this when the `reqType` is not one of "batch", "replay", "retl", "import":

```go
payload = append([]byte(`{"batch":[`), payload...)
payload = append(payload, []byte(`]}`)...)
```

Where do we do that here? :point_up:

**Review by fracasula — CHANGES_REQUESTED** [review]

File: gateway/handle.go [hunk 2]
```diff
@@ -779,7 +777,45 @@ func (gw *Handle) extractJobsFromInternalBatchPayload(reqType string, body []byt
...
+			stat.Report(gw.stats)
+			return nil, errors.New(response.NotRudderEvent)
+		}
+		msgPayload, err = sjson.SetBytes(msgPayload, "rudderId", rudderId.String())
+		if err != nil {
+			stat.RequestFailed(response.NotRudderEvent)
+			stat.Report(gw.stats)
+			return nil, errors.New(response.NotRudderEvent)
+		}
+		err = gw.streamMsgValidator(&msg)
```
[the remark this answers is not in the document]
Comment (BonapartePC) [reply gateway/handle.go]: Done

File: gateway/handle.go [hunk 3]
```diff
@@ -779,7 +777,45 @@ func (gw *Handle) extractJobsFromInternalBatchPayload(reqType string, body []byt
 
 	for _, msg := range messages {
 		stat := gwstats.SourceStat{ReqType: reqType}
-		err := gw.streamMsgValidator(&msg)
+		msgPayload := msg.Payload
```
[the remark this answers is not in the document]
Comment (BonapartePC) [reply gateway/handle.go]: Done

File: gateway/gateway_test.go [hunk 4]
```diff
@@ -2241,6 +2217,141 @@ var _ = Describe("Gateway", func() {
...
+				} `json:"batch"`
+			}
+			err = json.Unmarshal(jobsWithStats[0].job.EventPayload, &job)
+			Expect(err).To(BeNil())
+			Expect(job.Batch).To(HaveLen(1))
+			Expect(job.Batch[0].MessageID).To(Not(BeEmpty()))
+			Expect(job.Batch[0].RudderID).To(Not(BeEmpty()))
+		})
+
+		It("doesn't override if messageID already exists in payload", func() {
```
[the remark this answers is not in the document]
Comment (BonapartePC) [reply gateway/gateway_test.go:2266]: No, only we add `rudderId`, even in the ingestion service, it's adding `rudderId` but not checking its existence.

File: gateway/gateway_test.go [hunk 5]
```diff
@@ -2241,6 +2217,141 @@ var _ = Describe("Gateway", func() {
...
+				} `json:"batch"`
+			}
+			jobForm := jobsWithStats[0].job
+			err = json.Unmarshal(jobForm.EventPayload, &job)
+			Expect(err).To(BeNil())
+			Expect(job.Batch).To(HaveLen(1))
+			Expect(job.Batch[0].MessageID).To(ContainSubstring("dummyMessageID"))
+		})
+
+		It("adds type and batch in the request payload if it's a non batch call in RequestType Properties", func() {
```
[the remark this answers is not in the document]
Comment (BonapartePC) [reply gateway/gateway_test.go]: This is checking the default case. Made this test more generic and added another test case for batch, replay, retl and import

File: gateway/gateway_test.go (hunk 1 above)
Comment (BonapartePC) [reply gateway/gateway_test.go:2331]: `batch` here is not the request type but the endpoint srcrouter is hitting on server. So, it's always batch for all the tests. In the ingestion service, we are adding the following code only for convenience and we are iterating over it and sending each event:
```
payload = append([]byte(`{"batch":[`), payload...)
payload = append(payload, []byte(`]}`)...)
```

**Review by ktgowtham — APPROVED** [review]

File: gateway/gateway_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (fracasula) [reply gateway/gateway_test.go:2266]: Yeah makes sense, it's also not important in case an old ingestion-svc sets it and we overwrite it here :+1:

**Pull request merged**